httpx/middleware.py

```diff
@@ -1,61 +1,91 @@
+import functools
 import typing
+from base64 import b64encode
 
-from ..config import DEFAULT_MAX_REDIRECTS, CertTypes, TimeoutTypes, VerifyTypes
-from ..exceptions import RedirectBodyUnavailable, RedirectLoop, TooManyRedirects
-from ..models import URL, AsyncRequest, AsyncResponse, Cookies, Headers
-from ..status_codes import codes
-from .base import AsyncDispatcher
+from .config import DEFAULT_MAX_REDIRECTS
+from .exceptions import RedirectBodyUnavailable, RedirectLoop, TooManyRedirects
+from .models import URL, AsyncRequest, AsyncResponse, Cookies, Headers
+from .status_codes import codes
 
+Responder = typing.Callable[[AsyncRequest], typing.Coroutine[None, None, AsyncResponse]]
+Middleware = typing.Callable[
+    [AsyncRequest, typing.Callable], typing.Coroutine[None, None, AsyncResponse]
+]
 
-class RedirectDispatcher(AsyncDispatcher):
+
+def basic_auth(
+    username: typing.Union[str, bytes], password: typing.Union[str, bytes]
+) -> Middleware:
+    if isinstance(username, str):
+        username = username.encode("latin1")
+
+    if isinstance(password, str):
+        password = password.encode("latin1")
+
+    userpass = b":".join((username, password))
+    token = b64encode(userpass).decode().strip()
+    print(username, password, userpass, token)
+    authorization_header = f"Basic {token}"
+
+    async def dispatch(request: AsyncRequest, get_response: Responder) -> AsyncResponse:
+        request.headers["Authorization"] = authorization_header
+        return await get_response(request)
+
+    return dispatch
+
+
+def custom_auth(auth: typing.Callable[[AsyncRequest], AsyncRequest]) -> Middleware:
+    async def dispatch(request: AsyncRequest, get_response: Responder) -> AsyncResponse:
+        request = auth(request)
+        return await get_response(request)
+
+    return dispatch
+
+
+def redirect(
+    allow_redirects: bool = True,
+    max_redirects: int = DEFAULT_MAX_REDIRECTS,
+    cookies: typing.Optional[Cookies] = None,
+) -> Middleware:
+    return Redirect(
+        allow_redirects=allow_redirects, max_redirects=max_redirects, cookies=cookies
+    ).dispatch
+
+
+class Redirect:
```

Comment:
While I like the idea of middlewares being functional I somehow prefer them to be Dispatcher subclasses. While working on the Digest auth I realized there's quite a bit of state and logic that is fairly intertwined with dispatching.

That being said I do feel there's a bit of awkwardness in the subclassing, somehow I feel we're missing an abstaction but if I had to choose I'd rather have middlewares as dispatchers. 

Again I'd love to hear other people's thoughts on this. 🙂

Reply:
Thanks a lot for your review!

In your experience, what could be potential blockers to implementing Digest Auth using this callable interface? I think Redirect shows that we can add arbitrarily complex logic by deferring the implementation to a callable class, provided we only need to interact at the response layer (and not the underlying .send()).

(In retrospect I’d refactor the ad-hoc .dispatch() method on Redirect into a .__call__() method — and perhaps *all* middleware should be written in this style to avoid a mix of functions and callable classes.)

Reply:
I wasn't suggesting it was a blocker, sorry I that's how it came across, I was simply pointing out I found the class-based approach a bit more straightforward to read. I'll take a look your recent commits and re-review. 🙂